Avoid unnecessary blob uploads

miminar opened the pull request:
In an environment, where multiple users work with one daemon, build images and push them to authenticated repositories, the cross-mount usually fails because of permission issues.

Here's an example:
1. user _bob_ builds `reg.local/bob/php-base` image based on `docker.io/php` and pushes it to his private repository
   - Docker daemon remembers assocations between blobs and repositories:
     
       P <-> docker.io/php            (layer of php base image)
       P <-> reg.local/bob/php-base   (layer of php base image)
       N1<-> reg.local/bob/php-base   (layer on top of base image)
2. user _alice_ builds its own `reg.local/alice/php-base` image based on `docker.io/php` and pushes it to her private repository
   1. there's a new **N1** layer on top of **P**
   2. daemon attempts to mount **P** from `reg.local/bob/php-base` and fails because _alice_ cannot access _bob_'s private repositories
   3. the association between **P** and `reg.local/bob/php-base` is removed as a consequence
   4. both layers **P** and **N2** get fully uploaded
   5. association between **P** and `reg.local/alice/php-base` is remembered
3. so far so good, all works as expected
4. but now _bob_ adds one additional layer on top of `reg.local/bob/php-base` to build `reg.local/bob/myapp` and pushes it
   1. there's a new **M** layer on top of **N1**
   2. during a push, daemon does not attempt to stat the **P** because there's no association between **P** and `reg.local/bob/php-base` (it was removed in step 2)
   3. so it attempts to mount the blob again, this time from `reg.local/alice/php-base` private repository and fails
   4. the association between **P** and `reg.local/alice/php-base` is removed as a consequence
   5. blob **N1** is checked before upload, so it doesn't get re-uploaded
   6. the other blobs (**P** and **M**) get fully uploaded (while **P** already exists in the target repository)
   7. the association between **P** and `reg.local/bob/php-base` is remembered once again

This is a big problem on a large scale. In OpenShift, with tens or hundreds of users using the same daemon and a small set of base images, the cross repo mount succeeds very rarely. And due to associations being forgotten, Docker daemon mostly re-uploads all the blobs to the target repository where they already exist.

This PR attempts to address the above with following:
1. add an `AuthConfigHash` attribute to v2 metadata structure
2. forgett blob <-> repo association only if its `AuthConfigHash` matches the current one or it is yet unset
3. blob is checked for presence in target repository in all cases
4. blob mount is attempted several times until:
   - a successful mount
   - a maximum number of attempts is reached

The drawback of the 4th item is that it creates redundant blob uploads which are being cancelled shortly afterwards. If the cancellation fails, it relies on the fact that most registries do periodic upload cleanups. The point of it is to raise the chance of hitting a repository the user is authorized to pull from.

Maybe there's even a better solution. Your ideas and suggestions are very welcome.

miminar commented:
/cc @aaronlehmann, @BrianBland, @dmcgowan

aaronlehmann commented:
> before forgetting blob <-> repo association, it's first verified that the blob really doesn't exist or whether the current user just doesn't have the right permissions

I'd be careful with this. I think that right now, the registry will distinguish between bad permissions and the nonexistence of the blob, but arguably this leaks information and the registry should return 404s in both cases (this is what GitHub does when logged-out users try to access private repositories, for example). So I'm not sure it's a good idea to design around the registry's current behavior.

Could the problem be solved by storing a hash of the `AuthConfig` that was used to push or pull the blob in the `V2MetadataService`? If a record with a matching `AuthConfig` hash is used, we would prefer that one. Otherwise, if we end up re-pushing, we would store a new record with the current hash.

dmcgowan commented:
I agree with Aaron about relying on the registry to get information about the existence of the blob. We have tried to design the registry to not give this out this information, if it is possible then I think that is a bug.

> Could the problem be solved by storing a hash of the AuthConfig that was used to push or pull the blob in the V2MetadataService?

I agree with storing additional information on the engine side to make the decision on which to prefer and whether to remove from the list. We could probably just do it based on registry username (token hash if we don't have it) and only remove entries when attempting to mount with the same user. The metadata looks easily extensible and would rather explore that option first.

miminar commented:
Thank you guys for the input. Storing hash `AuthConfig` in the metadata sounds promising. I need to run some tests first to verify. I'll update the PR once done.

Commit:
Different number of retries for layers of different sizes

Classify blobs into three categories based on size.
Use a very limited number of mount attempts and no existence check for
small blobs. Use more attempts for bigger blobs.

Also remember blob associations during layer existence check.

Blob digests are now checked in the target repository from newest to
latest. If the blob exists and the metadata entry does not, it will be
created. If the blob is not found, the metadata entry will be removed.

Signed-off-by: Michal Minář <miminar@redhat.com>

Commit:
Try to cross-repo mount until success

Don't fallback back immediately to blob upload if the cross-repo mount
fails and layer upload is initiated by registry.

Instead cancel the upload and re-try cross-repo mount from different
source repository before doing full re-upload.

Signed-off-by: Michal Minář <miminar@redhat.com>

Commit:
Compare V2Metadata with associated auth config

to avoid unnecessary blob re-uploads.

Cross repository mount from particular repo will most probably fail if
the user pushing to the registry is not the same as the one who pulled
or pushed to the source repo.

This PR attempts first to cross-repo mount from the source repositories
associated with the pusher's auth config. Then it falls back to other
repositories sorted from the most similar to the target repo to the
least.

It also prevents metadata deletion in cases where cross-repo mount fails
and the auth config hashes differ.

Signed-off-by: Michal Minář <miminar@redhat.com>

miminar commented:
> Could the problem be solved by storing a hash of the AuthConfig that was used to push or pull the blob in the V2MetadataService? If a record with a matching AuthConfig hash is used, we would prefer that one. Otherwise, if we end up re-pushing, we would store a new record with the current hash.

@aaronlehmann Works great as expected! I've split the commit into smaller, edible pieces.

aaronlehmann commented on distribution/push_v2.go:
```diff
@@ -454,5 +515,108 @@ func layerAlreadyExists(ctx context.Context, metadata []metadata.V2Metadata, rep
+	if bla.arr[i].AuthConfigHash == bla.authConfigHash && bla.arr[j].AuthConfigHash != bla.authConfigHash {
+		return true
+	}
+	if (bla.arr[i].AuthConfigHash == bla.authConfigHash) == (bla.arr[j].AuthConfigHash == bla.authConfigHash) {
```
```go
if bla.arr[i].AuthConfigHash != bla.authConfigHash && bla.arr[j].AuthConfigHash == bla.authConfigHash {
        return false
}
```

miminar replied:
Indeed looks much better. Thanks!

aaronlehmann commented on distribution/push_v2.go:
```diff
@@ -291,86 +302,141 @@ func (pd *v2PushDescriptor) Upload(ctx context.Context, progressOutput progress.
+	mounted bool,
+	layerUpload distribution.BlobWriter,
+	err error,
+) {
```
Four return values feels excessive. Since this is only called in one place, it might be better to leave the code in-line instead of splitting it out into this function.

miminar replied:
Inlined.

aaronlehmann commented on distribution/push_v2.go:
```diff
@@ -454,5 +515,108 @@ func layerAlreadyExists(ctx context.Context, metadata []metadata.V2Metadata, rep
+
+func cancelLayerUpload(ctx context.Context, dgst digest.Digest, layerUpload distribution.BlobWriter) {
+	if layerUpload != nil {
+		go func() {
```
Could you explain why this is done in a goroutine? It feels like it would be a little cleaner to cancel the previous attempt before starting a new one. Does the `Cancel` function add significant latency?

miminar replied:
I just wanted to speed up cross-repo mounting. I don't have a strong argument for doing this concurrently. I just felt that this would be a good optimization.

aaronlehmann commented on distribution/metadata/v2_metadata_service.go:
```diff
@@ -15,10 +18,24 @@ type V2MetadataService struct {
+	if authConfig == nil {
+		return "", nil
+	}
+	buf, err := json.Marshal(authConfig)
```
I'm wondering if we should marshal a copy of `authConfig` that has the password removed. Two reasons come to mind:

1. Paranoia. If the password is low-complexity, someone might be able to reverse it using the hash of `authConfig`.
2. Currently, password changes would cause a hash mismatch. But we probably don't want them to.

dmcgowan replied:
This was partially why I suggested just using the username. Another is that updates to the auth config structure could cause the hash to change.

aaronlehmann replied:
I think something along the following lines would also work:

```go
hashInput := authConfig.Username
if authConfig.IdentityToken != "" {
        hashInput = authConfig.IdentityToken
} else if authConfig.RegistryToken != "" {
        hashInput = authConfig.RegistryToken
}
return digest.FromBytes([]byte(hashInput)).String(), nil
```

It's still important to do the hashing, because we don't want to put bare tokens in the distribution metadata store.

miminar replied:
Unfortunately, using this wouldn't help OpenShift. We use dummy user name almost everywhere. We differentiate based on the password which is a bearer token passed to OpenShift API which determines user's identity.

We could address it on our side though if you are strongly against storing the password.

aaronlehmann replied:
I think I'm okay with including the password in the hash input, but want to take care that we handle anything including passwords sensitively. One suggestion would be to store a HMAC of the record data using a hash of `AuthConfig` as the key. For example:

```go
	authConfigDigest := digest.FromBytes([]byte(hashInput))
	mac := hmac.New(sha256.New, []byte(authConfigDigest))
	mac.Write([]byte(v2Metadata.LayerDigest))
	mac.Write([]byte(v2Metadata.SourceRepository))
	expectedMAC := mac.Sum(nil)
	if hmac.Equal(v2Metadata.StoredMAC, expectedMAC) {
		// This AuthConfig is the same one the record was created with
	}
```

In this approach, the other fields of `V2Metadata` are covered by the HMAC, so they effectively provide a salt for the sensitive information we're hashing from the `AuthConfig`. It would be very difficult for someone to extract useful information about `AuthConfig` from the result because the details of the record are hashed in. It would prevent attacks where someone checks hashes against a table of hash results pregenerated from known username/password pairs, for example.

miminar replied:
I like it! Rewritten as suggested.

aaronlehmann reviewed (changes requested):
Design LGTM

dmcgowan commented on distribution/push_v2.go:
```diff
@@ -291,86 +302,141 @@ func (pd *v2PushDescriptor) Upload(ctx context.Context, progressOutput progress.
-		layerUpload, err = bs.Create(ctx)
+		return err.Descriptor, true, layerUpload, nil
+	}
+	if len(mountFrom.SourceRepository) > 0 && (authConfigHash == mountFrom.AuthConfigHash || len(mountFrom.AuthConfigHash) == 0) {
```
nit: above checked as `mountFrom.SourceRepository != ""`, here as `len(mountFrom.SourceRepository) > 0`

miminar replied:
replaced

Commit:
Allow to mock V2MetadataService in unit tests

Signed-off-by: Michal Minář <miminar@redhat.com>

miminar commented on distribution/metadata/v2_metadata_service.go:
```diff
@@ -15,10 +18,24 @@ type V2MetadataService struct {
+		return "", err
+	}
+	sum := sha256.Sum256(buf)
+	return hex.EncodeToString(sum[:]), nil
```
Good point, rewritten as suggested.Good point, rewritten as suggested.

aaronlehmann commented:
It looks like the unit test needs to be updated:

```
16:19:25 --- FAIL: TestGetRepositoryMountCandidates (0.00s)
16:19:25 	push_v2_test.go:354: [sort more items] got unexpected number of candidates: 5 != 6
16:19:25 	push_v2_test.go:358: [sort more items] candidate 1 does not match expected: metadata.V2Metadata{AuthConfigHash:"hash", Digest:"9", SourceRepository:"127.0.0.1/foo/app"} != metadata.V2Metadata{AuthConfigHash:"efgh", Digest:"7", SourceRepository:"127.0.0.1/foo/bar"}
16:19:25 	push_v2_test.go:358: [sort more items] candidate 2 does not match expected: metadata.V2Metadata{AuthConfigHash:"hash", Digest:"5", SourceRepository:"127.0.0.1/foo"} != metadata.V2Metadata{AuthConfigHash:"hash", Digest:"9", SourceRepository:"127.0.0.1/foo/app"}
16:19:25 	push_v2_test.go:358: [sort more items] candidate 3 does not match expected: metadata.V2Metadata{AuthConfigHash:"hash", Digest:"6", SourceRepository:"127.0.0.1/bar"} != metadata.V2Metadata{AuthConfigHash:"hash", Digest:"5", SourceRepository:"127.0.0.1/foo"}
16:19:25 	push_v2_test.go:358: [sort more items] candidate 4 does not match expected: metadata.V2Metadata{AuthConfigHash:"efgh", Digest:"2", SourceRepository:"127.0.0.1/hello-world"} != metadata.V2Metadata{AuthConfigHash:"hash", Digest:"6", SourceRepository:"127.0.0.1/bar"}
16:19:25 	push_v2_test.go:362: [sort more items] missing expected candidate at position 5 (metadata.V2Metadata{AuthConfigHash:"efgh", Digest:"2", SourceRepository:"127.0.0.1/hello-world"})
16:19:25 FAIL
```

I see that you updated the code not to hash the password. Were you able to handle this on your side?

miminar commented:
> It looks like the unit test needs to be updated:

Yes, sorry about that. Updated.

> I see that you updated the code not to hash the password. Were you able to handle this on your side?

I've switched to the hmac as suggested. It works very well. Hope it is up to your expectations.

Handling it on our side would be a quest for long winter nights.

aaronlehmann commented on distribution/metadata/v2_metadata_service.go:
```diff
@@ -1,52 +1,130 @@
 	SourceRepository string
+	// Hmac hashes above attributes with recent authconfig digest used as a key in order to determine matching
+	// metadata entries accompanied by the same credentials without actually exposing them.
+	Hmac string
```
Nit: can we call this `HMAC`, and uppercase `HMAC` in the function names below?

miminar replied:
Renamed.

aaronlehmann commented:
Looks good overall. Thanks for being so responsive to my suggestions.

The only thing I'm wondering about now is the change to `layerAlreadyExists`. It looks like it will now try to stat up to a few digests from other repositories, whereas before it only tried digests that were associated with the current repository. I'm curious about the motivation behind this. I would think that it would be better to go straight to a cross-repository mount when we don't know that the digest exists in the current repository, so we don't pay a latency penalty if the digest turns out not to exist in the current repository.

miminar commented:
> The only thing I'm wondering about now is the change to layerAlreadyExists. It looks like it will now try to stat up to a few digests from other repositories, whereas before it only tried digests that were associated with the current repository. I'm curious about the motivation behind this. I would think that it would be better to go straight to a cross-repository mount when we don't know that the digest exists in the current repository, so we don't pay a latency penalty if the digest turns out not to exist in the current repository.

My reasoning is that when blob exists in the target repo but no mapping exists yet for the target registry, the blob gets re-uploaded. Similarly when the mapping for the target repository does not exist, but there's a plenty of mapping to other repositories in the same registry that the user cannot access. Re-uploading of blobs existing in target repository doesn't seem right.

Nevertheless these cases will be less common. So I'm fine with keeping the existence checks only for metadata entries matching the target repo.

aaronlehmann commented:
> My reasoning is that when blob exists in the target repo but no mapping exists yet for the target registry, the blob gets re-uploaded. Similarly when the mapping for the target repository does not exist, but there's a plenty of mapping to other repositories in the same registry that the user cannot access. Re-uploading of blobs existing in target repository doesn't seem right.

> Nevertheless these cases will be less common. So I'm fine with keeping the existence checks only for metadata entries matching the target repo.

Yeah, there isn't really a clear right or wrong answer. But here's a few thoughts:

- Many (most?) layers are very small, so re-uploading is not necessarily worse than doing speculative existence checks.

- If we check for existence of digests that we don't have any mapping for in the repository, I think it's better to do that after trying cross-repository mounts, because the mounts have a higher probability of succeeding (at least in the general case - I realize OpenShift may be different).

If we're willing to take on extra complexity, we might consider making the steps depend on the layer size. For example, a layer under 100 KB could attempt a single cross-repository mount, and fall back to just uploading the layer. Layers between 100 KB and 5 MB could try more than one cross-repository mounts before falling back to uploading. And layers larger than 5 MB could try a limited number of cross-repository mounts, then try statting a few blob digests in the target repository, and if none of this works, do the upload.

I just made these numbers up so you could easily make a case for different numbers. Unfortunately no set of heuristics will suit all cases well. With a fast CPU and network connection, uploading even large blobs will be faster than checking for several blobs that turn out to not exist. With a slow CPU or network, it can be worth doing many checks if it avoids the need to re-upload.

Another thought is to parallelize these existence checks, since the real issue is latency. But again, that comes at the cost of complexity.

The simple option would be to only do existence checks when the repository in the metadata matches the target repository.

miminar commented:
I'm more inclined to your layer size heuristic for determining number of mount/existence attempts. It doesn't add too much complexity since the loops are already there, it just turns constants into variables.

Making the concurrent checks would be more complex and I don't think it's worth it since all the layers are already pushed in parallel. With above heuristics, in the worst case scenario, the smaller layers get pushed fast with very little latency and by that time, existence checks for the big layers may be already done, so the channel will be utilized from the beginning to the end.

I'll update the PR accordingly. Thanks for the input.

miminar commented:
Updated. Using suggested blob size heuristics for choosing a number of mount and existence check attempts.

aaronlehmann commented:
Thanks.

I think it makes sense to check at least one mapping in the target repository before attempting cross-repository mounts. The check in the target repository is more likely to succeed, and the output will  be less confusing ("Layer already exists" instead of "Mounted from <some random repository>") when pushing an image twice in a row.

miminar commented:
> I think it makes sense to check at least one mapping in the target repository before attempting cross-repository mounts. The check in the target repository is more likely to succeed, and the output will be less confusing ("Layer already exists" instead of "Mounted from <some random repository>") when pushing an image twice in a row.

I agree. Would you rather do all the existence checks before the first mount attempt or do just one at first, then try the cross mounts and if that fails, attempt few more existence checks for the big blobs?

You're right about the cross-repo mount having higher success rate in general. So making just one check at first, then attempting cross mounts, then 1 or two additional existence checks for big blobs should improve success rate for initial queries. The complexity will be a bit higher though.

aaronlehmann commented:
I'd vote for this ordering:

1. Existence check of one or two digests that are associated with the target repository in our metadata (I think one is probably enough)
2. A few cross repository mounts, depending on the size of the blob
3. For big blobs, speculative existence checks in the target repository, for digests that aren't associated with the target repository

miminar commented:
> I'd vote for this ordering:

Agreed. Code updated. Thanks!

aaronlehmann commented on distribution/push_v2.go:
```diff
@@ -348,21 +356,44 @@ func (pd *v2PushDescriptor) Upload(ctx context.Context, progressOutput progress.
+
+	if maxExistenceChecks-len(pd.checkedDigests) > 0 {
+		// do additional layer existence checks with other known digests if any
+		descriptor, exists, err := pd.layerAlreadyExists(ctx, progressOutput, diffID, checkOtherRepositories, maxExistenceChecks-len(pd.checkedDigests), v2Metadata)
```
Will this second call to `layerAlreadyExist` duplicate the single in-repository check that was done above? I notice that `v2Metadata` is not changed between the two calls.

miminar replied:
No, there's `(*v2PushDescriptor).checkedDigests` keeping all the processed digests to avoid redundant checks. See https://github.com/docker/docker/pull/26564/commits/9f05002b2dd247a6f4b3891ac030ec5104d17f10#diff-5482853d8bae37e26e93f996db0c3cecR515

aaronlehmann commented on distribution/push_v2.go:
```diff
@@ -404,55 +458,246 @@ func (pd *v2PushDescriptor) Upload(ctx context.Context, progressOutput progress.
+		meta := &candidates[i]
+		if _, exists := digestToMetadata[meta.Digest]; exists {
+			// keep reference just to the first mapping (the best mount candidate)
+			continue
```
Seems like this `continue`, and the one below, should not count towards the checks limit (i.e. the `i++` should be at the end of the for loop body.

miminar replied:
It does not, the `i` is not compared against check limits. Nevertheless I broke up the for-loop condition to make it more readable.

aaronlehmann reviewed (approved):
LGTM

dmcgowan commented on distribution/push_v2.go:
```diff
@@ -404,55 +458,249 @@ func (pd *v2PushDescriptor) Upload(ctx context.Context, progressOutput progress.
+	pathComponents []string
+	// added[i] stands for an order of addition of item arr[i]
+	// the higher the number, the newer the entry
+	added []int
```
Could this be achieved by reversing the original array and using [stable sort](https://golang.org/pkg/sort/#Stable)

miminar replied:
Sure it could. Good catch! Resolved.

dmcgowan reviewed (commented):
One question about the sorting then LGTM

dmcgowan commented on distribution/push_v2.go:
```diff
@@ -404,55 +458,237 @@ func (pd *v2PushDescriptor) Upload(ctx context.Context, progressOutput progress.
+func sortV2MetadataByLikenessAndAge(repoInfo reference.Named, hmacKey []byte, marr []metadata.V2Metadata) {
+	// reverse the metadata array to shift the newest entries to the beginning
+	for i := 0; i < len(marr)/2; i++ {
+		marr[i], marr[len(marr)-i-1] = marr[len(marr)-i-1], marr[i]
```
This reversal should only be run if `len(marr) > 1`

aaronlehmann replied:
If `len(marr) < 1`, won't the condition in the for loop never be satisfied?

dmcgowan replied:
yes, you are right

dmcgowan commented:
LGTM

thaJeztah commented:
@dmcgowan @aaronlehmann are any documentation changes needed for this change?

aaronlehmann commented:
I don't think this has any docs impacts.

thaJeztah reviewed (approved):
ok, it if looks good to @dmcgowan, it LGTM

thaJeztah commented:
experimental failure is a flaky test, but restarting for good measure

dmcgowan commented:
@thaJeztah wouldn't let me :+1: your approval comment :smile: 

But I agree with Aaron, I don't think there is anything in the docs related to cross repository push since it is more an implementation detail and optimization we try to do invisibly to the user

thaJeztah commented:
@dmcgowan yes, I thought it would be, just double checking. Anyway; thanks for reviewing this, and moving this forward; and of course thanks @miminar for contributing. This looks like a really nice improvement!

miminar commented:
It was a pleasure to contribute. Thanks for all the valuable feedback and review. I hope it will avail not only OpenShift users.

thaJeztah commented:
All green now - merging
